src/resolve/resolved-dns-scope.c

```diff
@@ -388,7 +388,15 @@ static int dns_scope_socket(
         if (s->link) {
                 be32_t ifindex_be = htobe32(ifindex);
 
-                if (sa.sa.sa_family == AF_INET) {
+                if (sa.sa.sa_family == AF_INET && type == SOCK_DGRAM) {
+                        /* Send the request on the link associated to the server */
+                        char ifname[IF_NAMESIZE + 1] = {0};
+                        if_indextoname(ifindex, ifname);
+                        log_debug("Send the request on the interface %d / %s", ifindex, ifname);
+                        r = setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, ifname, strlen(ifname));
+                        if (r < 0)
+                                return -errno;
+                } else if (sa.sa.sa_family == AF_INET) {
                         r = setsockopt(fd, IPPROTO_IP, IP_UNICAST_IF, &ifindex_be, sizeof(ifindex_be));
```

Comment:
`IP_UNICAST_IF` has no effect on anything but IPv4 UDP. I would remove this. We don't know what effect it might have on stream sockets (and similar) in the future.